include/picongpu/plugins/openPMD/writer/ParticleAttribute.hpp

```diff
@@ -0,0 +1,157 @@
+
+            OpenPMDName< T_Identifier > openPMDName;
+            ::openPMD::Record record = particleSpecies[ openPMDName() ];
+            ::openPMD::Datatype openPMDType =
+                ::openPMD::determineDatatype< ComponentType >();
+
+            // get the SI scaling, dimensionality and weighting of the attribute
+            OpenPMDUnit< T_Identifier > openPMDUnit;
+            std::vector< float_64 > unit = openPMDUnit();
+            OpenPMDUnitDimension< T_Identifier > openPMDUnitDimension;
+            std::vector< float_64 > unitDimension = openPMDUnitDimension();
+            const bool macroWeightedBool = MacroWeighted< T_Identifier >::get();
+            const uint32_t macroWeighted = ( macroWeightedBool ? 1 : 0 );
+            const float_64 weightingPower =
+                WeightingPower< T_Identifier >::get();
+
+            PMACC_ASSERT(
+                unit.size() == components ); // unitSI for each component
+            PMACC_ASSERT(
+                unitDimension.size() == 7 ); // seven openPMD base units
+
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD:  (begin) write species attribute: %1%" ) %
+                Identifier::getName();
+
+            std::shared_ptr< ComponentType > storeBfr;
+            if ( elements > 0 )
+                storeBfr = std::shared_ptr< ComponentType >{
+                    new ComponentType[ elements ],
+                    []( ComponentType * ptr ) { delete[] ptr; }
+                };
+
+            for( uint32_t d = 0; d < components; d++ )
+            {
+                ::openPMD::RecordComponent recordComponent = components > 1
+                    ? record[ name_lookup[ d ] ]
+                    : record[::openPMD::MeshRecordComponent::SCALAR ];
+                ::openPMD::Datatype openPMDType =
+                    ::openPMD::determineDatatype< ComponentType >();
+
+                ValueType * dataPtr = frame.getIdentifier( Identifier() )
+                                          .getPointer(); // can be moved up?
+                auto storePtr = storeBfr.get();
+
+/* copy strided data from source to temporary buffer */
+#pragma omp parallel for simd
+                for( size_t i = 0; i < elements; ++i )
+                {
+                    storePtr[ i ] = reinterpret_cast< ComponentType * >(
+                        dataPtr )[ d + i * components ];
+                }
+
+                params
+                    ->initDataset< DIM1 >(
+                        recordComponent,
+                        openPMDType,
+                        { globalElements },
+                        true,
+                        params->compressionMethod );
+                if( storeBfr )
+                    recordComponent.storeChunk(
+                        storeBfr, { globalOffset }, { elements } );
+
+                if( unit.size() >= ( d + 1 ) )
+                {
+                    recordComponent.setUnitSI( unit.at( d ) );
```

Comment:
please use `unit[ d ]` since you already guard for avoiding out of memory access within the if condition.